### PR TITLE
[SPARK-21040][CORE] Speculate tasks which are running on decommission executors

### PR DESCRIPTION
**Pull request by prakharjain09**
### What changes were proposed in this pull request?
This PR adds functionality to consider the running tasks on decommission executors based on some config.
In spark-on-cloud , we sometimes already know that an executor won't be alive for more than fix amount of time. Ex- In AWS Spot nodes, once we get the notification, we know that a node will be gone in 120 seconds.
So if the running tasks on the decommissioning executors may run beyond currentTime+120 seconds, then they are candidate for speculation.

### Why are the changes needed?
Currently when an executor is decommission, we stop scheduling new tasks on those executors but the already running tasks keeps on running on them. Based on the cloud, we might know beforehand that an executor won't be alive for more than a preconfigured time. Different cloud providers gives different timeouts before they take away the nodes. For Ex- In case of AWS spot nodes, an executor won't be alive for more than 120 seconds. We can utilize this information in cloud environments and take better decisions about speculating the already running tasks on decommission executors.

### Does this PR introduce _any_ user-facing change?
Yes. This PR adds a new config "spark.executor.decommission.killInterval" which they can explicitly set based on the cloud environment where they are running.

### How was this patch tested?
Added UT.

### CODE REVIEW

**Comment by prakharjain09** [comment]
I am adding more UTs for this change. So added WIP in title.

@holdenk Please give some feedback on the actual logic.

**Comment by HyukjinKwon** [comment]
cc @ScrapCodes, @vanzin, @jiangxb1987, @Ngone51 too

**Comment by holdenk** [comment]
Jenkins ok to test.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1123]: nit:

```suggestion
    executorDecommissionKillInterval.foreach { interval =>
      val executorKillTime = clock.getTimeMillis() + interval
      taskInfos.filter(info.executorId == execId && info.running).foreach { info =>
        tidToExecutorKillTimeMapping(into.taskId) = executorKillTime
      }
    }
```

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1853]: is it too small since your given examples are at least using minute?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1955]: nit: `Noting` -> `Nothing`

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1996]: Do we need to apply resourceOffer for exec2 since it's decommissioning?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1926]: It seems `startedTasks` is updated asynchronically by `FakeDAGScheduler`. Maybe we could use `eventually` to pretend the potential failure.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1944]: Could you assert `tidToExecutorKillTimeMapping` here?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1973]: > Task 4 has completed 16 ms

Do you mean task 4 has been running for 16ms?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1057]: We need to remove the corresponding task from `tidToExecutorKillTimeMapping` if `speculated` is true?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1973]: BTW, it should be TASK 2 instead of TASK 4?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:2000]: I think this test maybe not necessary because the above test has already covered it:

```
    // Task2 started at t=0, so it can still finish before t=15 (Median task runtime = 10)
    // Task3 started at t=6, so it might not finish before t=15. So Task 3 should be part
    // of speculativeTasks
```

?

**Comment by Ngone51** [comment]
BTW, what do we do for those tasks if speculative is disabled? Waiting for executor lost action?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1123]: 1. Added `executorDecommissionKillInterval.foreach { interval =>` to remove the if check.
2. Its better to filter runningTasksSet as compared to taskInfos as it is going to be relatively small.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1955]: done.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1996]: Yes - we should not apply to exec2. Removed it.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1926]: `dagScheduler.taskStarted` is called as part of `resourceOffer` and the overridden taskStarted method of FakeDagScheduler appends it to `startedTasks` set. So it seems synchronus only?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1944]: done.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1973]: Yea I meant Task2. Sorry for the confusion - Updated the line to - `At t=16 ms, Task 2 has been running for 16 ms`.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1057]: We are removing entries from tidToExecutorKillTimeMapping at time when it is removed from `runningTasksSet`.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:2000]: Agreed. The above tests checks this functionality also. Should I remove this?

**Comment by prakharjain09** [comment]
@Ngone51  Thanks for the review.

> BTW, what do we do for those tasks if speculative is disabled? Waiting for executor lost action?
If speculation is disabled, either the running tasks on those executors will succeed or they might fail and will be retried when the executor is taken away forcefully (due to node loss etc).

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1926]: I see. you're right.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1057]: Then, is it possible that we add redundant speculate tasks for the same tid?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:2001]: Yes, please.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1057]: The `copiesRunning(index) == 1` condition inside `checkAndSubmitSpeculatableTask` will prevent it. Irrespective of executor decommissioning, that check prevents that only 1 speculatable task can run for each task.

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1057]: I mean if it's possible to add redundant speculate tasks to `speculatableTasks` rather than running multiple  speculate tasks for the same tid at the same time.

I think the actual prevention is `!speculatableTasks.contains(index)` in `checkAndSubmitSpeculatableTask`, but never mind since `speculatableTasks` is a `HashSet`.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: done.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Converted to TimeUnit.Seconds and updated the tests accordingly.

**Comment by prakharjain09** [comment]
@Ngone51 I have addressed all the review comments. Can you please take a look at this PR?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1942]: nit: redundant whitespace

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:1988]: Can we unify the description for the task? As far as I see, we have `TASK 2`, `TASK2`, `Task2`, `task index 2` in the test.

**Review by Ngone51 — COMMENTED** [review]
LGTM expect 2 minor comments, cc @holdenk @jiangxb1987

**Comment by prakharjain09** [comment]
@holdenk @cloud-fan Please review the changes.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: removed.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (prakharjain09) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: done. Now using `TASK 2` everywhere in test.

**Comment by prakharjain09** [comment]
@holdenk @jiangxb1987 Please review the changes.

> LGTM expect 2 minor comments, cc @holdenk @jiangxb1987

**Comment by Ngone51** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1849]: Nit: does the Ex in the "go down after decommissioning signal Ex- around 2 mins" stands for "example"?
Sorry I have not seen this abbreviation used before and still not sure it exists. What about "i.e." that even used in Spark documentation several times?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1851]: Nit: Missing dot from the end of sentence.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1050]: I am just curious why this solution (introducing the kill time per task ID map: `tidToExecutorKillTimeMapping`) is chosen over storing the kill time per executor (`executorToKillTimeMapping` or something like that). In the later case here would be  something like: 

```scala
if (!speculated && executorToKillTimeMapping.nonEmpty) {
  val taskInfo = taskInfos(tid)
  executorToKillTimeMapping.get(taskInfo.executorId).foreach {  executorKillTime =>
  ...
  }
}
```

**Review by attilapiros — APPROVED** [review]
Few comments/question but the changes LGTM.

One more question: the `checkAndSubmitSpeculatableTask` checks whether the number of running copies from the task is one:
https://github.com/apache/spark/blob/4affa58f95f893ef6de1c1bf1c6b731468a2519d/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L1007

It could happen that we already have two running copies of the task but both are running on decommissioning executors (i.o on the same host). I am not sure whether this is worth to be considered (I mean here or maybe in a followup PR). What do you think?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:1050]: Yes - We can store executorToKillTimeMapping also. In that case, we have to clean the map when executor is removed. For that we may have to make changes to pass "removeExecutor" signal till TaskSetManager.

I didn't see much difference in both the approaches as total entries we will be creating for a given executor in current approach will be equal to total tasks running for that executor (which shouldn't be a lot).

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: changed to "i.e.".

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: fixed.

**Comment by prakharjain09** [comment]
@attilapiros Thanks for the review.

> One more question: the checkAndSubmitSpeculatableTask checks whether the number of running copies from the task is one:
spark/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Line 1007 in 4affa58
 if (!successful(index) && copiesRunning(index) == 1 && 
> It could happen that we already have two running copies of the task but both are running on decommissioning executors (i.o on the same host). I am not sure whether this is worth to be considered (I mean here or maybe in a followup PR). What do you think?

Yes it is possible that two copies can be running on decommissioning executors and so we will end up not speculating for it because of above checks. I felt it might be a corner case which we can take as separate/follow-up PR.

**Comment by prakharjain09** [comment]
@holdenk @cloud-fan @Dooyoung-Hwang Please review the changes.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1850]: So the timeout is decided by the cloud vendors? What does this config specify?

Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1850]: @cloud-fan This config can be set by users based on their setups. If they are using AWS spot nodes, timeout can be set to somewhere around 120 seconds, if they are using fix duration 6hrs spot blocks (say they decommission executors at 5:45), timeout can be set to 15 mins and so on.

If user doesn't set this timeout, things will remain as they were and tasks running on decommission executors  won't get any special treatment with respect to speculation.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1850]: is it possible that Spark can get this timeout value from the cluster manager? So that users don't need to set it manually. cc @holdenk

Comment (prakharjain09) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1850]: @cloud-fan As per my understanding, Worker Decommissioning is getting triggered currently using SIGPWR signal (and not via some message coming from YARN/Kubernetes Cluster manager). So getting this timeout from Spark Cluster Manager might not be possible. We might be able to do this once Spark's Worker Decommissioning logic starts triggering via communication from YARN etc in future. cc  @holdenk

**Comment by prakharjain09** [comment]
@holdenk @cloud-fan @Dooyoung-Hwang Please review the changes.

**Comment by holdenk** [comment]
I'll be holding off on this while we wait for the SPIP.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1850]: I believe there are some situations where we can know the length of time from the cluster manager or from Spark it's self, but not all. I think having a configurable default for folks who know their cloud provider environment makes sense

**Comment by holdenk** [comment]
Took a quick look, thanks for working on this. I think having a timeout to kill the executors regardless (e.g. a max decommissioning time) and the speculation are both useful. I'll follow up more once we've decided on the design in OSS.

**Comment by prakharjain09** [comment]
@holdenk Please review the changes.

**Comment by holdenk** [comment]
This looks good to me, I'm excited for us to have this improvement. Thanks for working on this and thanks to all the folks who took the time to review it. Sorry I have not been as active with the reviews quite as much as I want (new puppy joined our household last week).

**Comment by holdenk** [comment]
Merged to the development branch!

**Comment by jiangxb1987** [comment]
shouldn't we trigger another round of jenkins test before merging this?

**Comment by holdenk** [comment]
This chunk of code is pretty standalone and theres no conflicts so I didn't feel it necessary, it's a judgement call and Jenkins is in the middle of an upgrade.

**Pull request closed without merging**
